Lf 4765 create get endpoint to fetch list of i ps

Duncan-Brain opened the pull request:
**Description**

** REBASED - in case you pulled locally already

~~Waiting for LF-4767 to merge so I can delete frontend mock and add after_date param to query function.~~

This implements the backend endpoint to get irrigation prescriptions from esci.

It also adds tests for this endpoint.

Notes: 
- Sorry about the commits they will be hard to read, it got too messy for my liking:
  - I looked into farm local date because we don't have an accurate one for DST (everything is UTC based but should be IANA timezone code based is my understanding), thanks Joyce for pointing me in the direction of using our existing browser local pattern despite it not being quite correct for farm local. (one of these days it would be nice to get dates consistent and correct ...but not today)
  - Also got confused with the other ip route and if I should be putting this route there
  - should have reverted date commits instead of deleteing manually, it led to some back pakage-lock later on needing correcting -- sorry
- Went ham on full model types, I have been lazy with this before now, also moved some types around as I thought they were broadly useful, let me know if its no good, but I would ask to keep it.

Jira link: [LF-4765](https://lite-farm.atlassian.net/browse/LF-4765)

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Passes test case
- [x] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [x] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files


[LF-4765]: https://lite-farm.atlassian.net/browse/LF-4765?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Duncan-Brain commented on packages/api/src/controllers/irrigationPrescriptionController.ts:
```diff
@@ -0,0 +1,100 @@
+            throw partnerErrors.shift();
+          }
+          return res.status(200).send(irrigationPrescriptions);
+        } else {
```
To be deleted...

Duncan-Brain commented on packages/api/src/controllers/irrigationPrescriptionRequestController.ts:
```diff
@@ -13,28 +13,26 @@
   shouldSend?: string;
 }
 
-export interface LiteFarmRequest extends Request<unknown, unknown, unknown, LiteFarmQuery> {
```
Wanted to reuse this so made a generic.

Duncan-Brain commented on packages/api/src/models/irrigationTypesModel.js:
```diff
@@ -17,7 +17,7 @@ class IrrigationTypesModel extends BaseModel {
       required: [''],
       properties: {
-        irrigation_type_id: { type: 'string' },
+        irrigation_type_id: { type: 'integer' },
```
Error in model setup

Duncan-Brain commented on packages/api/src/models/types.ts:
```diff
@@ -0,0 +1,596 @@
+ * Once models can be converted to TS merge with the model file.
+ *
+ */
+
```
Half the lines of code changes are here. I would consider ignoring reviewing the types file, because if they aren't working it will not build! The system/rules I used to make the file are maybe of interest.

I typed everything I touched during this PR, I have found myself making a lot of placeholder types with just an id property and it seemed like a waste so I decided to just do it finally.

Duncan-Brain commented on packages/api/tests/mock.factories.js:
```diff
@@ -2548,13 +2548,16 @@ async function animal_type_use_relationshipFactory({
     return knex('addon_partner')
       .insert({
-        ...partner,
+        ...(partner ? partner : fakePartner),
```
I need to be able to set the id property to work with ESci specific brand and id.

Duncan-Brain replied:
Yeah it is hardcoded in that endpoint. I like adding it your esci specific function but I think I was looking at two reasons why I didn't.

- I didn't want to mess up other tests -- this is backwards compatible 
- If I add it there I would want to return the `addonPartner` also from it -- so I believed I was keeping it true to the functions role as a simple wrapper around `farm_addonFactory()`

So I could add it to connectFarmToEnsemble but then I would want to return `ESCIAddonPartner` or hardcode the id in the test to compare to the result. I dno.. if its okay to return the addonPartner from this function too I can rework it?

kathyavini replied:
1. It won't mess up the other tests; they aren't concerned with the id belonging to ESci
2. What do you need the addon partner for rather thant the farm addon? Your code is only using the farmAddon at present

kathyavini replied:
The other tests don't care about what the ESci partner id is, so they wouldn't be affected (they can take a `1` as easily as any other ID). You can leave it as is though, now that you're passing and asserting on the EsciAddonPartner -- that wasn't part of the test when I initially reviewed!

Duncan-Brain commented on packages/api/tests/utils/date.ts:
```diff
@@ -0,0 +1,50 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
```
Brought here from the frontend -- would be nice to use `/shared` more. Should I?

Duncan-Brain replied:
Decided not to do it here, moving forward I will remember you are good with this.

Duncan-Brain commented on packages/api/tests/utils/ensembleUtils.ts:
```diff
@@ -15,16 +15,89 @@
   return { farmAddon };
 };
+
+type fakeIrrigationPrescriptionsProps = {
```
This isn't currently like our other "fakes" I would like to fix that but maybe once it is not doing both beta and tests mocking.

Duncan-Brain commented on packages/api/tests/utils/ensembleUtils.ts:
```diff
@@ -15,16 +15,89 @@
+ * @param farm_id - The ID of the farm to retrieve mock data for.
+ * @returns A promise that resolves to formatted irrigation prescription data.
+ */
+export const fakeIrrigationPrescriptions = async ({
```
This isn't currently like our other "fakes" I would like to fix that but maybe once it is not doing both beta and tests mocking.

Duncan-Brain replied:
I was working on the assumption task completion flow was not yet testable. Let me see what I can remove now that it is merged. Yes the external id on task would have had to be manually inserted in db or just view the null values.

Granted I shouldn't have tried to merge the mocks probably. Let me see what can be improved.

Duncan-Brain replied:
I think this should now be addressed. Sorry the tests were bad indeed. I am so used to just doing something to show it kind of works and moving on. I have a lot of trouble with our tests...

The test now gets prescriptions before linking task, then creates a task with it linked, refetches the prescriptions with tasks to show they all succeed.

Duncan-Brain commented on packages/webapp/src/containers/saga.js:
```diff
@@ -631,6 +631,7 @@ export function* clearOldFarmStateSaga() {
       'CustomAnimalTypes',
       'DefaultAnimalTypes', // needs to be cleared for KPI count
       'FarmAddon',
+      'IrrigationPrescriptions',
```
Did not know about this - TODO for me make Jira ticket to investigate if everything is up to date here or could import tags instead.

Duncan-Brain replied:
https://lite-farm.atlassian.net/browse/LF-4820

Duncan-Brain commented on packages/webapp/src/store/api/apiSlice.ts:
```diff
@@ -295,7 +296,15 @@ export const api = createApi({
     }),
     getIrrigationPrescriptions: build.query<IrrigationPrescription[], void>({
-      query: () => `${irrigationPrescriptionsUrl}`,
+      query: () => {
```
Updating params from previous pr

kathyavini commented on packages/api/tests/utils/ensembleUtils.ts:
```diff
@@ -15,16 +15,89 @@
+ */
+export const fakeIrrigationPrescriptions = async ({
+  farmId,
+  prescriptionIds = [1, 2],
```
We're going to need more than 2 Irrigation Prescription IDs to get a full test of the feature on beta. To get exactly one new IP ID each day (needed for notifications) [I used `.getUTCDay()`]((https://github.com/LiteFarmOrg/LiteFarm/pull/3762/files#diff-9b4bb7e42b9cbb676a613576cb9b7e2c0773777b3680979164ab7e8037abb897)) in the mock IP list there.

Thinking about it now, maybe what would be even better for this purpose is `.getUTCDate()`? Then we could move through 30 or 31 values (instead of 7) before duplicating, and it would be a pretty close approximation of the real ESci endpoint -- what do you think?

Duncan-Brain replied:
🙏

kathyavini commented on packages/api/tests/utils/testDataSetup.ts:
```diff
@@ -56,7 +45,10 @@ export async function returnUserFarms(role: number) {
  */
 export async function setupFarmEnvironment(role: number = 1) {
-  const { mainFarm: farm, user: owner } = await returnUserFarms(1);
+  const { mainFarm: farm, user: owner } = (await returnUserFarms(1)) as {
```
You see a TS error here in your environment that requires type casting?

I don't see any ts errors with the original code so I can't test this, but what about typing the return of `returnUserFarms` instead of casting:

```ts
export async function returnUserFarms(
  role: number
): Promise<{ mainFarm: Farm; user: User }> {
```

Duncan-Brain replied:
I have no clue but you are absolutely right!

Duncan-Brain replied:
Fixed!

kathyavini commented on packages/webapp/src/store/api/apiSlice.ts:
```diff
@@ -295,7 +296,15 @@ export const api = createApi({
+        const params = new URLSearchParams({ startTime, endTime });
+
+        return `${irrigationPrescriptionUrl}?${params.toString()}`;
+      },
```
I have a hunch (could be tested live once the mock is fixed up a bit) that we'll also need to do the reverse of the below and trigger this GET when creating an irrigation task.

Duncan-Brain replied:
Yes thanks so much 🙏  I think you are right.

Duncan-Brain replied:
Added this to create and delete.

kathyavini reviewed (commented):
Thanks for getting started with a lot of great patterns here ❤️ The model types will be super useful and I love the endpoint genererics setup! And hopefully someday we really will expand beyond ESci partners and some dev somewhere will appreciate your unwavering commitment to extensibility! Even though it's not implemented anywhere else, it's nice to have the pattern to refer to if/when (🤞) it happens.

The main concern for me is just the timing and the scope of the mock, which I expanded on [here](https://github.com/LiteFarmOrg/LiteFarm/pull/3755/files#r2082345115). Right now the mock is short circuiting the entire controller, which means on beta we can't get eyes and testing on the part of your controller that e.g. associates the IPs with the right existing tasks. Since we have a working flow from IP > task so we want to test, rather than short-circuit that.

Ideally, I think we would only want to be replacing the ESci return. So that would mean only pulling a valid crop-supporting location -- I'll probably copy the logic in GET details to make sure we're always grabbing the same one -- and then optionally a management plan on that location (or not), and then dates + appropriate ID. We could then test the rest of the controller logic AND the whole flow from the UX perspective, on beta.

Duncan-Brain commented on packages/api/src/controllers/irrigationPrescriptionController.ts:
```diff
@@ -0,0 +1,100 @@
+          // Loop through addon partners
+          for (const farmAddonPartnerId of farmAddonPartnerIds) {
+            try {
+              const addonPartner = PARTNER_ID_MAP[farmAddonPartnerId.addon_partner_id.toString()];
```
It didn't like numeric keys, let me try again.

Duncan-Brain replied:
easy fixed, I must have started with enum or something

Duncan-Brain commented on packages/api/src/controllers/irrigationPrescriptionController.ts:
```diff
@@ -0,0 +1,100 @@
+
+          // Return an error if there are no prescriptions
+          if (!irrigationPrescriptions.length && partnerErrors.length) {
+            throw partnerErrors.shift();
```
Yup good point, it does not make sense at all. I might have started with an error for Partner_id = 0 and thought I should dump that or something but that doesnt make sense either. Thanks!

Duncan-Brain commented on packages/api/src/util/ensembleService.ts:
```diff
@@ -23,12 +23,110 @@ import LocationModel from '../models/locationModel.js';
+import { AddonPartner, Farm, FarmAddon } from '../models/types.js';
+
+/**
+ * Retrieves the external organisation IDs for a specific farm and partner.
```
I swear I double checked all these 😞 .. thanks!

Duncan-Brain commented on packages/api/src/util/ensembleService.ts:
```diff
@@ -23,12 +23,110 @@ import LocationModel from '../models/locationModel.js';
+  const addonPartnerId = await getAddonPartnerId();
+  const externalOrganizationIds = await getExternalOrganisationIds(farmId, addonPartnerId);
+
+  // Endpoint config
```
I did not notice it as a pattern ahah,  in `ensemble.js` we have `unclaimSensor()` which seems to do stuff in the function with the first `ensembleApiCall()` . And in this `ensembleService.ts` file the `ensembleAPICall()` has no return values... so there would be nothing to do after anyways!

Most of our other endpoints the 'ESci' part is not encapsulated, so thats all I was doing here. But I can easily make it into two functions.

Let me know if you think that would be more readable!

Duncan-Brain replied:
I see it now, formatting data is generic to all partners 👍  will fix

Duncan-Brain replied:
Fixed as part of encapsulating comment for Anto.

Duncan-Brain commented on packages/api/src/util/ensembleService.ts:
```diff
@@ -23,12 +23,110 @@ import LocationModel from '../models/locationModel.js';
+ */
+export const getIrrigationPrescriptions = async (
+  farmId: string,
+  startTime?: string,
```
I was thinking from ESci's standpoint it is probably fine to define one, both or neither. If I was Morten I would probably build that way -- but we can refine in the future based on what he gives us.

So maybe I forgot to restrict it from our side in the controller so that we definitely get what we want from our side.

Duncan-Brain replied:
I have added a middleware check to restrict it from our endpoint side of things for now if thats ok

Duncan-Brain commented on packages/api/src/util/ensembleService.ts:
```diff
@@ -23,12 +23,110 @@ import LocationModel from '../models/locationModel.js';
+    return [];
+  }
+
+  if (!isExternalIrrigationPrescriptionArray(data)) {
```
Thants an AI assist for sure.

Duncan-Brain commented on packages/api/src/util/ensembleService.types.ts:
```diff
@@ -83,3 +80,44 @@ export interface EnsembleLocationAndCropData {
+): data is ExternalIrrigationPrescription[] {
+  return (
+    Array.isArray(data) &&
+    data.every((item): item is IrrigationPrescription => {
```
Looks like I forgot to update when I split the type 🙏

Duncan-Brain replied:
Should be fixed

Duncan-Brain commented on packages/api/src/util/ensembleService.ts:
```diff
@@ -23,12 +23,110 @@ import LocationModel from '../models/locationModel.js';
+ * Retrieves the external organisation IDs for a specific farm and partner.
+ *
+ * @param farm_id - The ID of the farm to retrieve external organisation IDs for.
+ * @returns A promise that resolves to the organisation IDs for the given farm and partner.
```
Should be all better 🙏

Duncan-Brain commented on packages/api/src/util/ensembleService.ts:
```diff
@@ -23,12 +23,110 @@ import LocationModel from '../models/locationModel.js';
+};
+
+/**
+ * Returns a list of mocked prescriptions based on a specific farm_id.
```
Should be all beter 🙏

Duncan-Brain commented:
@kathyavini I was working on the basis that the Tasks flow was not ready yet -- just the migration. I had a good chunk already in draft before the Tasks PR went through. And it was merged while I was putting finishing touches before opening for review. Thats why it Tasks was not taken into account. ... Let me see if I can address your concerns now.

antsgar reviewed (commented):
Sorry this took me a while to review! I had started it the other day and had to switch to the Soil breakdown. I've added a few comments, but it's overall looking good -- I'll re-review once you've addressed Joyce's comment ☺️

Duncan-Brain commented on packages/api/tests/irrigation_prescription.test.ts:
```diff
@@ -0,0 +1,159 @@
+
+describe('Get Irrigation Prescription Tests', () => {
+  let ESciAddonPartner: AddonPartner;
+  async function postIrrigationTask({
```
I hate the factory functions haha. They are just so not the flow of things, theres a whole graph to populate and in order to do it like the factory way I had to add three factories irrigation task, task, location task and I am sure they don't all have the right properties (sometimes useful to check against).

It is now done with factories and it looks okay. I still hate the pattern lol

kathyavini replied:
Ah sorry, I should have been a lot more specific, as I think this is actually a few levels beyond what I was thinking!

I just meant you included an endpoint to POST the irrigation task (the POST endpoint that would be normally only be defined in `tasks.test.js`) instead of calling `mocks.taskFactory()` -- only that part was what I was wondering about! That location tasks graph we have to create is indeed pretty messy, so I do think that's totally reason enough to not want to create it manually, and maybe just a comment to that end for similar situations?

Since you already did all the hard work of creating that task graph with factories, let's definitely keep that logic, but what about:
- renaming those function to something like 'buildExternalIrrigationPrescription()` since in our codebase `*Factory` functions are always for populating the database via `knex` insert, which these do not do
- moving the file to `tests/utils/ensemble.ts` or a dedicated file

Since IPs aren't database entities they really don't belong in that mocks file or with that `*Factory` naming convention. They're perfectly fine mocks / builders, but just not really factories as we use that term, so if they were moved and renamed you wouldn't have to make that disclaimer comment about their abnormality. What do you think?

Duncan-Brain replied:
It is similar to a factory in the sense that it uses other factories, and it would create something if we were the owners of that db. But I get your point about the naming convention... and it being a different construct because we don't have many tests for external endpoints. 

Uhmm.. it is also kind of like a fake* because it is not-retrievable afterwards... but it needs non static data. 

It is also not necessarily ESCI specific beyond this line : `id: id ?? 1`.

Can I just rename it as you suggested with `build*` and keep it here in mocks? It really feels like a hybrid mock for external endpoints (conforming to LiteFarms desired data structure) and this is where we keep all other mocks.

kathyavini replied:
@Duncan-Brain that works for me! My main concern is people less familiar with the feature assuming we have a database entity for IP based on there being a `Factory` function, so as long as there is no factory in the name that's already a big improvement. I see what you mean about it being hybrid.

Duncan-Brain replied:
K renamed.

Duncan-Brain commented on packages/webapp/src/apiConfig.js:
```diff
@@ -92,7 +92,7 @@ export const soilAmendmentPurposesUrl = `${URI}/soil_amendment_purposes`;
 export const productUrl = `${URI}/product`;
 export const farmAddonUrl = `${URI}/farm_addon`;
-export const irrigationPrescriptionsUrl = `${URI}/irrigation_prescriptions`;
+export const irrigationPrescriptionUrl = `${URI}/irrigation_prescription`;
```
sure... will update the jira ticket description then

Duncan-Brain commented on packages/api/src/controllers/irrigationPrescriptionController.ts:
```diff
@@ -0,0 +1,100 @@
+
+        if (shouldSend === 'true') {
+          // Check for registered farm addons (only esci for now)
+          // @ts-expect-error - farm_id is guaranteed here by the checkScope middleware with single argument
+          const farmAddonPartnerIds = await FarmAddonModel.getDistinctFarmAddonPartnerIds(farm_id);
```
I like having a comment for unclear assertions. Assertions should be used sparingly.

I have a test branch for a specific fix for this open mentioned with Joyces comment: https://github.com/LiteFarmOrg/LiteFarm/pull/3755#discussion_r2091295599

antsgar replied:
I'd say ts-expect-error should be used even more sparingly, and my understanding is that it's meant to be used temporarily. If you're absolutely sure `farm_id` would be present I think the non-null assertion would be okay. I did look at the other branch, and although I like the idea of adding a type guard I honestly feel it overcomplicates the code with not that much added benefit. But I'm curious to know what others think too

antsgar replied:
This is the type guard I'm referring to https://github.com/LiteFarmOrg/LiteFarm/pull/3772/files#diff-31ffdd44c889de461e56828613f90ff13ba191db724cda9bdb7b6e90e8937d48R45-R52.

A type assertion is as bad as a non-null assertion (arguably worse?) so my feeling is the balance between having a +185 -42 diff to end up with a solution that isn't ideal vs achieving a solution that isn't good either but just uses one character doesn't add up. I'd be totally okay with keeping the @ts-expect-error as a non-ideal solution too, that's not a blocker for me!

Duncan-Brain replied:
Great, I would like to keep the ts-expect error then since it has a comment explaining its existence and we may not know the best way to type it. 

Thanks for sharing the type guard,  I think the one you are pointing too has nothing to do with the farm_id, the query param type guard check has been moved to middleware in this PR but in the draft PR it is outdated.

To be fair the actual code change to assert the correct type is like ~ +20, the rest is changing a core file to typescript and adding more types which is generally useful. I am really not saying we should do it but it seems like we are striving to have mostly correct typings -- I personally think this solution is possibly the most correct we can get with express.

To continue the best practice discussion -- because I like chatting about it 😄 :
>  A type assertion is as bad as a non-null assertion (arguably worse?)

Its not that either asserting that is necessarily bad, it is asserting out of context or to use as a temporary solution that is bad. There are genuine cases for each but being extra careful about both is important I think. For this case I think we would be using it here as a temporary solution, out of context, that could lead to serious code reuse errors.

The farm_id type guard that already exists in checkScope is two files away from the controller. If it were right beside it and not working in context (like the TS example, and other examples on internet I can find) I would agree non-null assertion.

Type casting assertions are also useful when maybe you cannot type check like external package or endpoint.

Duncan-Brain commented on packages/api/src/controllers/irrigationPrescriptionController.ts:
```diff
@@ -0,0 +1,100 @@
+          return res.status(200).send(irrigationPrescriptions);
+        } else {
+          // Return data for dev purposes + QA
+          const mockData = await fakeIrrigationPrescriptions({
```
This only affects the Esci endpoint now. Let me know if it works for you.

Duncan-Brain commented on packages/api/src/controllers/irrigationPrescriptionController.ts:
```diff
@@ -0,0 +1,100 @@
+const PARTNER_ID_MAP: Record<string, Partial<AddonFunctions>> = {
+  '0': {},
+  '1': ESciAddon,
+};
```
Made a service file.

Duncan-Brain commented:
I think I have addressed all the comments. Just noting that when I tested the full flow including creating a task I think some parts of task flow may not be correct:
- Wrong `external_prescription_id` possibly being sent to backend
- Not able to create a irrigation task using the same `external_prescription_id` between farms -- should this not be farm specific? (even though it will should likely never happen)

kathyavini commented:
> Just noting that when I tested the full flow including creating a task I think some parts of task flow may not be correct:
>
> Wrong external_prescription_id possibly being sent to backend
> Not able to create a irrigation task using the same external_prescription_id between farms -- should this not be farm specific? (even though it will should likely never happen)

@Duncan-Brain did you maybe not pull the most recent version of integration before you merged it? Those adjusts were merged in #3762 but I don't see that code on this branch.

kathyavini commented:
@Duncan-Brain actually, is your local code definitely the same as what you have pushed to this branch? Can you double-check? You have a line here:

```ts
return shouldSend ? ESciAddon : Mocks;
```

that controls when the mocking happens, but `shouldSend` is not a boolean on this branch -- it's still the string `false` -- so this is always hitting the ESci endpoint and returning a 404. The result is no IPs in the frontend, so if you are seeing them, something is not aligned... 🤔 Can we check on this tomorrow morning before/after Sprint Planning?

Duncan-Brain commented:
@kathyavini Thanks for merging integration, yes I think you are right. 

Type and conditional is updated. Must have had my frontend cache why I didn't see it 🙏 . Damned attention to detail haha

kathyavini commented:
Anto gave her verbal approval for this one (and we'll be moving to her review not blocking) and it looks good to me!
